canonicalwebteam/discourse_docs/app.py

```diff
@@ -1,26 +1,81 @@
-from canonicalwebteam.discourse_docs.blueprint import build_blueprint
+import flask
+from requests.exceptions import HTTPError
+
+from canonicalwebteam.discourse_docs.exceptions import PathNotFoundError
+from canonicalwebteam.discourse_docs.parsers import (
+    parse_topic,
+    parse_index,
+    resolve_path,
+)
 
 
 class DiscourseDocs(object):
+    """
+    A Flask extension object to serve documentation pages,
+    pulling the documentation content from Discourse.
+
+    :param api: A DiscourseAPI for retrieving Discourse topics
+    :param index_topic_id: ID of a forum topic containing nav & URL map
+    :param category_id: Only show docs from topics in this forum category
+    :param url_prefix: URL prefix for hosting under (Default: /docs)
+    :param document_template: Path to a template for docs pages
+                              (Default: docs/document.html)
+    """
+
     def __init__(
         self,
-        app=None,
-        model=None,
+        api,
+        index_topic_id,
+        category_id,
         url_prefix="/docs",
         document_template="docs/document.html",
     ):
-        self.app = app
-        if app is not None:
-            self.init_app(app, model, url_prefix)
+        self.url_prefix = url_prefix
+        self.blueprint = flask.Blueprint("discourse_docs", __name__)
```

Comment:
I felt this was clearer, and I'll lay out my thinking below. Hopefully I can convince you =)

As we discussed in our 1 to 1, I think wrapping it in a `generate_blueprint` function is an unnecessary level of encapsulation - it would make this init function do so little as to be meaningless to read, and instead make the reader jump one level further down to see the meat of what's going on.

The interface for this module is `DiscourseDocs().init_app(app)`. That's how this module is intended to be used. If we wished to expose `generate_blueprint()` for consumption by the client, I would argue, why are we bothering with the `DiscourseDocs` extension object at all? We might as well just tell clients to do `app.register_blueprint(generate_blueprint, url_prefix)` themselves. But instead we're following the [recommended Flask extension structure](http://flask.pocoo.org/docs/1.0/extensiondev/), but the entirety of the meat of the extension is the blueprint, so for all intents and purposes, the `DiscourseDocs` extension object is your `generate_blueprint` factory already.

And if, at some point in the future, it becomes useful to have the blueprint generation abstracted away for use elsewhere, then it can easily be encapsulated at that point. But until then it feels to me like [gold plating](https://blog.codinghorror.com/gold-plating/) (see also, [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)).

Do you think that's a fair characterisation?

Reply:
Yup, makes sense. 

I would still keep `DiscourseDocs().init_app(app)`.

But the idea is to keep the functions in a different part of memory, not on the discourse class:

```
import flask
from requests.exceptions import HTTPError

from canonicalwebteam.discourse_docs.exceptions import PathNotFoundError
from canonicalwebteam.discourse_docs.parsers import (
    parse_topic,
    parse_index,
    resolve_path,
)

 def generate_blueprint(api,):
       blueprint = flask.Blueprint("discourse_docs", __name__)

        blueprint.route("/")
        blueprint.route("/<path:path>")
        def document(path=""):
            """
            A Flask view function to serve
            topics pulled from Discourse as documentation pages.
            """

            # Ensure path has a leading slash
            path = "/" + path.lstrip("/")

            index = parse_index(api.get_topic(index_topic_id))

            if path == "/":
                document = index
            else:
                try:
                    topic_id = resolve_path(path)
                except PathNotFoundError:
                    return flask.abort(404)

                if topic_id == index_topic_id:
                    return flask.redirect("/")

                try:
                    topic = api.get_topic(topic_id)
                except HTTPError as http_error:
                    return flask.abort(http_error.response.status_code)

                if category_id and topic["category_id"] != category_id:
                    return flask.abort(404)

                document = parse_topic(topic)

                if document["topic_path"] != path:
                    return flask.redirect(document["topic_path"])

            return flask.render_template(
                document_template,
                document=document,
                navigation=index["navigation"],
                forum_url=api.base_url,
            )

class DiscourseDocs(object):
    """
    A Flask extension object to serve documentation pages,
    pulling the documentation content from Discourse.
    :param api: A DiscourseAPI for retrieving Discourse topics
    :param index_topic_id: ID of a forum topic containing nav & URL map
    :param category_id: Only show docs from topics in this forum category
    :param url_prefix: URL prefix for hosting under (Default: /docs)
    :param document_template: Path to a template for docs pages
                              (Default: docs/document.html)
    """

    def __init__(
        self,
        api,
        index_topic_id,
        category_id,
        url_prefix="/docs",
        document_template="docs/document.html",
    ):
        self.url_prefix = url_prefix
   
  
    def init_app(self, app):
       blueprint = generate_blueprint(api, all other params)
       app.register_blueprint(blueprint, url_prefix=self.url_prefix)
```

Reply:
It ultimately does the same thing. So not a blocker. I just feel that it is easier to discern the different parameters. 

No need to rewrite though. Its your code and works fine :)

Reply:
I think I did understand what you meant - the example you give is what I thought you meant... but why?

> No new level of abstraction is introduced, as it is still the same file.

Even though it's in the same file, it is a new level of abstraction nonetheless. If I'm stepping through the code from my application where I've initialised `DiscourseDocs()` to see what's happening, I'm first going to land at `def __init__(`, read that, then keep following the tree. It's one more "step into" in a debugger, one more "go to declaration" in my IDE, or simply one more manual logical step to parse.

> possible extraction at a later point without doing it now ... if we ever extract the blueprint

This is the bit that I find to be gold-plating. This is not a requirement that we have now, we don't currently know if we'll ever have, and this restructure can very easily be done as soon as it's needed. The [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle, for me, says this future possibility should basically never be a reason for adding complexity right now. 

> at the same time we introduce function parameters

This bit has some value. However, these parameters are so similar to the set of paraemters being passed through to `def __init__` anyway that in this case I don't think encapsulation for the sake of this one bit of clarity is worth it. The parameters that `DiscourseDocs` expects are:

`api, index_topic_id, category_id, url_prefix, document_template`

The parameters that `generate_blueprint` will need are:

`api, index_topic_id, category_id, document_template`

The only one you avoid is `url_prefix`, and with the `generate_blueprint` structure, you end up having to just pass 4 parameters one more level down the chain.

I'm afraid I'm really not seeing the value.

Reply:
I could even make it so `url_prefix` is an argument on `init_app` rather than `__init__`. I think I might do this. This will mean that the `__init__` function is *literally* just creating the blueprint, making it literally no different from the `generate_blueprint` function.

Reply:
Yeah true. It just comes down to personal preference here. I find the function approach easier to maintain and test, since its boundaries do not collide with `class DiscourseDocs(object):`.

Reply:
I've moved the `url_prefix` parameter to `init_app`. Thanks for that, I think that's a better structure.

I don't see the boundaries point. If `DiscourseDocs.__init__` is literally equivalent to `generate_blueprint`, they can both be tested in exactly the same fashion.

The only downside I can still see is that `DiscourseDocs.__init__` is not as descriptive a function name as `generate_blueprint`. To me, avoiding one more level of encapsulation is more important than this name, but I can see that genuinely is a personal choice.